File: front/src/api/tasks.ts
import { z } from "zod";
import { AxiosInstance } from "axios";

import { ClipSchema } from "@/api/clips";
import { TagSchema } from "@/api/tags";
import { SimpleUserSchema } from "@/api/user";
import { NoteSchema, type NoteUpdate, NoteUpdateSchema } from "@/api/notes";

import { GetManySchema, Page } from "./common";

export type State = "assigned" | "completed" | "verified" | "rejected";

export const TaskTagSchema = z.object({
  id: z.number().int(),
  task_id: z.number().int(),
  tag_id: z.number().int(),
  tag: TagSchema,
});

export type TaskTag = z.infer<typeof TaskTagSchema>;

export const TaskTagFilter = z.object({
  task__eq: z.number().int().optional(),
  recording__eq: z.number().int().optional(),
  tag__eq: z.number().int().optional(),
  created_at__before: z.coerce.date().optional(),
  created_at__after: z.coerce.date().optional(),
  project__eq: z.number().int().optional(),
  search: z.string().optional(),
  key__eq: z.string().optional(),
  key__has: z.string().optional(),
  value__eq: z.string().optional(),
  value__has: z.string().optional(),
});

export type TaskTagFilter = z.infer<typeof TaskTagFilter>;

export const TaskTagPageSchema = Page(TaskTagSchema);

export type TaskTagPage = z.infer<typeof TaskTagPageSchema>;

export const GetTaskTagsQuerySchema = z.intersection(
  GetManySchema,
  TaskTagFilter,
);

export type GetTaskTagsQuery = z.infer<typeof GetTaskTagsQuerySchema>;

export const StatusBadgeCreateSchema = z.object({
  state: z.enum(["assigned", "completed", "verified", "rejected"]),
});

export type StatusBadgeCreate = z.infer<typeof StatusBadgeCreateSchema>;

export const StatusBadgeSchema = z.object({
  id: z.number(),
  state: z.enum(["assigned", "completed", "verified", "rejected"]),
  user: SimpleUserSchema,
});

export type StatusBadge = z.infer<typeof StatusBadgeSchema>;

export const TaskCreateSchema = z.object({
  clip_id: z.number(),
  project_id: z.number(),
});

export type TaskCreate = z.infer<typeof TaskCreateSchema>;

export const TaskSchema = z.object({
  id: z.number(),
  clip_id: z.number(),
  project_id: z.number(),
  clip: ClipSchema,
  tags: z.array(TaskTagSchema),
  status_badges: z.array(StatusBadgeSchema),
  notes: z.array(NoteSchema),
});

export type Task = z.infer<typeof TaskSchema>;

export const TaskPageSchema = Page(TaskSchema);

export type TaskPage = z.infer<typeof TaskPageSchema>;

export const TaskFilterSchema = z.object({
  dataset__eq: z.number().optional(),
  dataset__isin: z.array(z.number()).optional(),
  project__eq: z.number().optional(),
  project__isin: z.array(z.number()).optional(),
  recording_tag__eq: z.number().optional(),
  pending__eq: z.boolean().optional(),
  assigned__eq: z.boolean().optional(),
  verified__eq: z.boolean().optional(),
  rejected__eq: z.boolean().optional(),
  assigned_to__eq: z.string().uuid().optional(),
});

export type TaskFilter = z.infer<typeof TaskFilterSchema>;

export const GetTasksSchema = z.intersection(GetManySchema, TaskFilterSchema);

export type GetTasks = z.infer<typeof GetTasksSchema>;

export const TaskNoteSchema = z.object({
  task_id: z.number().int(),
  note_id: z.number().int(),
  note: NoteSchema,
});

export type TaskNote = z.infer<typeof TaskNoteSchema>;

export const TaskNoteFilter = z.object({
  note__eq: z.number().int().optional(),
  created_by__eq: z.string().uuid().optional(),
  project__eq: z.number().int().optional(),
  task_eq: z.number().int().optional(),
  is_issue__eq: z.boolean().optional(),
  message__eq: z.string().optional(),
  message__has: z.string().optional(),
  created_at__before: z.coerce.date().optional(),
  created_at__after: z.coerce.date().optional(),
});

export type TaskNoteFilter = z.infer<typeof TaskNoteFilter>;

export const TaskNotePageSchema = Page(TaskNoteSchema);

export type TaskNotePage = z.infer<typeof TaskNotePageSchema>;

export const GetTaskNotesQuerySchema = z.intersection(
  GetManySchema,
  TaskNoteFilter,
);

export type GetTaskNotesQuery = z.infer<typeof GetTaskNotesQuerySchema>;

const DEFAULT_ENDPOINTS = {
  createMany: "/api/v1/tasks/",
  getMany: "/api/v1/tasks/",
  get: "/api/v1/tasks/detail/",
  delete: "/api/v1/tasks/detail/",
  addNote: "/api/v1/tasks/detail/notes/",
  updateNote: "/api/v1/tasks/detail/notes/",
  removeNote: "/api/v1/tasks/detail/notes/",
  addTag: "/api/v1/tasks/detail/tags/",
  removeTag: "/api/v1/tasks/detail/tags/",
  addBadge: "/api/v1/tasks/detail/badges/",
  removeBadge: "/api/v1/tasks/detail/badges/",
  getNotes: "/api/v1/tasks/notes/",
  getTags: "/api/v1/tasks/tags/",
};

export function registerTasksApi(
  instance: AxiosInstance,
  endpoints: typeof DEFAULT_ENDPOINTS = DEFAULT_ENDPOINTS,
) {
  async function createMany(data: TaskCreate[]): Promise<Task[]> {
    const body = z.array(TaskCreateSchema).parse(data);
    const response = await instance.post(endpoints.createMany, body);
    return z.array(TaskSchema).parse(response.data);
  }

  async function getMany(query: GetTasks): Promise<TaskPage> {
    const params = GetTasksSchema.parse(query);
    const response = await instance.get(endpoints.getMany, { params });
    return TaskPageSchema.parse(response.data);
  }

  async function get(task_id: number): Promise<Task> {
    const params = { task_id };
    const response = await instance.get(endpoints.get, { params });
    return TaskSchema.parse(response.data);
  }

  async function delete_(task_id: number): Promise<Task> {
    const params = { task_id };
    const response = await instance.delete(endpoints.delete, { params });
    return TaskSchema.parse(response.data);
  }

  async function addNote({
    task_id,
    message,
    is_issue,
  }: {
    task_id: number;
    message: string;
    is_issue: boolean;
  }): Promise<Task> {
    const response = await instance.post(
      endpoints.addNote,
      {
        message,
        is_issue,
      },
      {
        params: {
          task_id,
        },
      },
    );
    return TaskSchema.parse(response.data);
  }

  async function updateNote({
    task_id,
    note_id,
    data,
  }: {
    task_id: number;
    note_id: number;
    data: NoteUpdate;
  }) {
    const body = NoteUpdateSchema.parse(data);
    const response = await instance.patch(endpoints.updateNote, body, {
      params: {
        task_id,
        note_id,
      },
    });
    return TaskSchema.parse(response.data);
  }

  async function removeNote({
    task_id,
    note_id,
  }: {
    task_id: number;
    note_id: number;
  }) {
    const response = await instance.delete(endpoints.removeNote, {
      params: {
        task_id,
        note_id,
      },
    });
    return TaskSchema.parse(response.data);
  }

  async function addBadge({
    task_id,
    state,
  }: {
    task_id: number;
    state: string;
  }) {
    const body = StatusBadgeCreateSchema.parse({ state });
    const response = await instance.post(endpoints.addBadge, body, {
      params: {
        task_id,
      },
    });
    return TaskSchema.parse(response.data);
  }

  async function removeBadge(task_id: number, badge_id: number) {
    const response = await instance.delete(endpoints.removeBadge, {
      params: {
        task_id,
        badge_id,
      },
    });
    return TaskSchema.parse(response.data);
  }

  async function addTag({
    task_id,
    tag_id,
  }: {
    task_id: number;
    tag_id: number;
  }) {
    const response = await instance.post(
      endpoints.addTag,
      {},
      {
        params: {
          task_id,
          tag_id,
        },
      },
    );
    return TaskSchema.parse(response.data);
  }

  async function removeTag({
    task_id,
    tag_id,
  }: {
    task_id: number;
    tag_id: number;
  }) {
    const response = await instance.delete(endpoints.removeTag, {
      params: {
        task_id,
        tag_id,
      },
    });
    return TaskSchema.parse(response.data);
  }

  async function getNotes(query: GetTaskNotesQuery) {
    const params = GetTaskNotesQuerySchema.parse(query);
    const { data } = await instance.get(endpoints.getNotes, { params });
    return TaskNotePageSchema.parse(data);
  }

  async function getTags(query: GetTaskTagsQuery) {
    const params = GetTaskTagsQuerySchema.parse(query);
    const { data } = await instance.get(endpoints.getTags, { params });
    return TaskTagPageSchema.parse(data);
  }

  return {
    createMany,
    getMany,
    get,
    delete: delete_,
    getNotes,
    getTags,
    addNote,
    updateNote,
    removeNote,
    addTag,
    removeTag,
    addBadge,
    removeBadge,
  };
}

File: src/whombat/routes/tasks.py
"""REST API routes for annotation tasks."""
from fastapi import APIRouter, Depends, Body
from typing import Annotated

from whombat import api, schemas
from whombat.dependencies import ActiveUser, Session
from whombat.filters.task_notes import TaskNoteFilter
from whombat.filters.tasks import TaskFilter
from whombat.filters.task_tags import TaskTagFilter
from whombat.routes.types import Limit, Offset
from whombat.models.task import TaskState

__all__ = [
    "tasks_router",
]


tasks_router = APIRouter()


@tasks_router.post(
    "/",
    response_model=list[schemas.Task],
)
async def create_tasks(
    session: Session,
    data: list[schemas.TaskCreate],
):
    """Create multiple annotation tasks."""
    tasks = await api.tasks.create_many(
        session,
        data,
    )
    await session.commit()
    return tasks


@tasks_router.get(
    "/",
    response_model=schemas.Page[schemas.Task],
)
async def get_tasks(
    session: Session,
    limit: Limit = 10,
    offset: Offset = 0,
    filter: TaskFilter = Depends(TaskFilter),  # type: ignore
    sort_by: str = "-created_at",
):
    """Get a page of annotation tasks."""
    tasks, total = await api.tasks.get_many(
        session,
        limit=limit,
        offset=offset,
        filters=[filter],
        sort_by=sort_by,
    )
    return schemas.Page(
        items=tasks,
        total=total,
        limit=limit,
        offset=offset,
    )


@tasks_router.delete(
    "/detail/",
    response_model=schemas.Task,
)
async def delete_task(
    session: Session,
    task_id: int,
):
    """Remove a clip from an annotation project."""
    task = await api.tasks.delete(session, task_id)
    await session.commit()
    return task


@tasks_router.post(
    "/detail/notes/",
    response_model=schemas.Task,
)
async def create_task_note(
    session: Session,
    task_id: int,
    data: schemas.NoteCreate,
    user: ActiveUser,
):
    """Create a note for an annotation task."""
    note = await api.notes.create(
        session,
        schemas.NotePostCreate(
            created_by_id=user.id,
            **data.model_dump(),
        ),
    )
    task = await api.tasks.add_note(
        session,
        task_id,
        note.id,
    )
    await session.commit()
    return task


@tasks_router.delete(
    "/detail/notes/",
    response_model=schemas.Task,
)
async def delete_task_note(
    session: Session,
    task_id: int,
    note_id: int,
):
    """Delete a note from an annotation task."""
    task = await api.tasks.remove_note(
        session,
        task_id,
        note_id,
    )
    await session.commit()
    return task


@tasks_router.get(
    "/notes/",
    response_model=schemas.Page[schemas.TaskNote],
)
async def get_task_notes(
    session: Session,
    limit: Limit = 10,
    offset: Offset = 0,
    sort_by: str = "-created_at",
    filter: TaskNoteFilter = Depends(TaskNoteFilter),  # type: ignore
):
    """Get a page of task notes."""
    notes, total = await api.tasks.get_notes(
        session,
        limit=limit,
        offset=offset,
        filters=[filter],
        sort_by=sort_by,
    )
    return schemas.Page(
        items=notes,
        total=total,
        offset=offset,
        limit=limit,
    )


@tasks_router.get(
    "/tags/",
    response_model=schemas.Page[schemas.TaskTag],
)
async def get_task_tags(
    session: Session,
    limit: Limit = 10,
    offset: Offset = 0,
    sort_by: str = "-created_at",
    filter: TaskTagFilter = Depends(TaskTagFilter),  # type: ignore
):
    """Get a page of task tags."""
    tags, total = await api.tasks.get_tags(
        session,
        limit=limit,
        offset=offset,
        filters=[filter],
        sort_by=sort_by,
    )
    return schemas.Page(
        items=tags,
        total=total,
        offset=offset,
        limit=limit,
    )


@tasks_router.get(
    "/detail/",
    response_model=schemas.Task,
)
async def get_task(
    session: Session,
    task_id: int,
):
    """Get an annotation task."""
    task = await api.tasks.get_by_id(session, task_id)
    return task


@tasks_router.post(
    "/detail/tags/",
    response_model=schemas.Task,
)
async def add_task_tag(
    session: Session,
    task_id: int,
    tag_id: int,
    user: ActiveUser,
):
    """Create a tag for an annotation task."""
    task = await api.tasks.add_tag(
        session,
        task_id,
        tag_id,
        user.id,
    )
    await session.commit()
    return task


@tasks_router.delete(
    "/detail/tags/",
    response_model=schemas.Task,
)
async def remove_task_tag(
    session: Session,
    task_id: int,
    tag_id: int,
):
    """Remove a tag from an annotation task."""
    task = await api.tasks.remove_tag(
        session,
        task_id,
        tag_id,
    )
    await session.commit()
    return task


@tasks_router.post(
    "/detail/notes/",
    response_model=schemas.Task,
)
async def add_task_note(
    session: Session,
    task_id: int,
    data: schemas.NoteCreate,
    user: ActiveUser,
):
    """Add a note to an annotation task."""
    note = await api.notes.create(
        session,
        schemas.NotePostCreate(
            created_by_id=user.id,
            **dict(data),
        ),
    )
    task = await api.tasks.add_note(
        session,
        task_id,
        note.id,
    )
    await session.commit()
    return task


@tasks_router.patch(
    "/detail/notes/",
    response_model=schemas.Task,
)
async def update_task_note(
    session: Session,
    task_id: int,
    note_id: int,
    data: schemas.NoteUpdate,
):
    """Update a note on an annotation task."""
    task = await api.tasks.update_note(
        session,
        task_id,
        note_id,
        data,
    )
    await session.commit()
    return task


@tasks_router.post(
    "/detail/badges/",
    response_model=schemas.Task,
)
async def add_task_badge(
    session: Session,
    task_id: int,
    state: Annotated[TaskState, Body(embed=True)],
    user: ActiveUser,
):
    """Add a badge to an annotation task."""
    task = await api.tasks.add_status_badge(
        session,
        task_id,
        user.id,
        state,
    )
    await session.commit()
    return task


@tasks_router.delete(
    "/detail/badges/",
    response_model=schemas.Task,
)
async def remove_task_badge(
    session: Session,
    task_id: int,
    badge_id: int,
):
    """Remove a badge from an annotation task."""
    task = await api.tasks.remove_status_badge(
        session,
        task_id,
        badge_id,
    )
    await session.commit()
    return task
